### PR TITLE
[#7] Create default top navigation bar

### LINKED ISSUE #7: As a user, I can see the default top navigation bar
## Why
When first opening the application, the users will be able to see the default top navigation bar.

## Acceptance Criteria
- [ ] Create the default native top navigation bar.
- [ ] Set its background color to green as below design for styling purpose (our theme will be green).
- [ ] Set the screen's title in the center in white color.

## Resources
- Sample top navigation bar UI layout:
Screen Shot 2021-08-05 at 17 46 34

### PR DESCRIPTION
**Pull request by markgravity**
Resolved #7 

## What happened

When first opening the application, the users will be able to see the default top navigation bar.

## Insight

- Add navigation bar modifier to handle navigation background and title color

## Proof Of Work

Screen Shot 2021-08-11 at 13 17 35

### CODE REVIEW

File: NimbleMedium/Sources/Presentation/Views/Screen.swift [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+        self.content = content()
+        self.title = title
+        self.titleColor = titleColor
+    }
+
+    var body: some View {
+        NavigationView {
+            content
+                .navigationBarTitle(
+                    title ?? "",
```
Comment (minhnimble) [inline NimbleMedium/Sources/Presentation/Views/Screen.swift:29]: Please create a global constant for empty string name `EMPTY_STRING` for the usage here and future usage, thanks 🙏

File: NimbleMedium/Sources/Presentation/Modules/Home/HomeView.swift [hunk 2]
```diff
@@ -8,13 +8,16 @@
 import SwiftUI
 
 struct HomeView: View {
+
     var body: some View {
-        Text(Localizable.homeTitle())
-            .padding()
+        Screen(title: Localizable.homeTitle()) {
+            Text("This is home")
```
Comment (minhnimble) [inline NimbleMedium/Sources/Presentation/Modules/Home/HomeView.swift:14]: Just an optional that we can add a `TODO` here noting that we will remove this `Text` at a later time, this is just for testing purpose only 😁

**Review by minhnimble — COMMENTED** [review]
Nice work, just a few minor comments

File: NimbleMedium/Sources/Constants/Constants.swift [hunk 3]
```diff
@@ -4,3 +4,6 @@
 //
 //  Created by Mark G on 10/08/2021.
 //
+
+// swiftlint:disable identifier_name
+let EMPTY_STRING = ""
```
Comment (minhnimble) [inline NimbleMedium/Sources/Constants/Constants.swift:9]: or we can go with Apple's best practice which is the camelCase version - `emptyString` 😁 disable lint at every place seems a bit annoying I think

Comment (suho) [reply NimbleMedium/Sources/Constants/Constants.swift:9]: I prefer
```swift
extension String {

   static let empty = ""
}
```

File: NimbleMedium/Sources/Presentation/Views/Screen.swift [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    private let titleColor: Color
+
+    // swiftlint:disable type_contents_order
+    init(title: String? = nil,
+         titleColor: Color = .white,
+         @ViewBuilder content: () -> Content) {
+        self.content = content()
+        self.title = title
+        self.titleColor = titleColor
+    }
```
Comment (suho) [inline NimbleMedium/Sources/Presentation/Views/Screen.swift:23]:
```suggestion
    init(
        title: String? = nil,
        titleColor: Color = .white,
        @ViewBuilder content: () -> Content
    ) {
        self.content = content()
        self.title = title
        self.titleColor = titleColor
    }
```

File: NimbleMedium/Sources/Constants/Constants.swift (hunk 3 above)
Comment (markgravity) [reply NimbleMedium/Sources/Constants/Constants.swift:9]: @minhnimble how about this? 🤔 
```swift
title.orEmpty()
```

Comment (minhnimble) [reply NimbleMedium/Sources/Constants/Constants.swift:9]: @markgravity can write an extension for checking if a string is empty `orEmpty()` as well 👍. But we still need an empty string constant to work with. Can use @suho for the constant declaration 🙏

Comment (markgravity) [reply NimbleMedium/Sources/Constants/Constants.swift:9]: Let me combine both  👍

File: NimbleMedium/Sources/Presentation/Views/Screen.swift [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+//  NimbleMedium
+//
+//  Created by Mark G on 10/08/2021.
+//
+
+import SwiftUI
+
+struct Screen<Content>: View where Content: View {
+
+    private let content: Content
```
Comment (suho) [inline NimbleMedium/Sources/Presentation/Views/Screen.swift:12]: I would prefer to define the property like the code below as we usually create and use the view in `body`
```suggestion
private let content: () -> Content
```

In some cases, we need to duplicate the content ([GridStack](https://www.hackingwithswift.com/books/ios-swiftui/custom-containers)), so don't need to create the view in init

File: NimbleMedium/Sources/Presentation/Views/Screen.swift [hunk 6]
```diff
@@ -0,0 +1,38 @@
+//
+//  Screen.swift
+//  NimbleMedium
+//
+//  Created by Mark G on 10/08/2021.
+//
+
+import SwiftUI
+
+struct Screen<Content>: View where Content: View {
```
Comment (suho) [inline NimbleMedium/Sources/Presentation/Views/Screen.swift:10]:
```suggestion
struct Screen<Content: View>: View {
```

File: NimbleMedium/Sources/Supports/Extensions/NavigationBarModifier.swift [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+        }
+    }
+}
+
+extension View {
+
+    func navigationBarColor(
+        backgroundColor: Color = .clear,
+        titleColor: Color = .white) -> some View {
+        self.modifier(
```
Comment (suho) [inline NimbleMedium/Sources/Supports/Extensions/NavigationBarModifier.swift:54]: Can we omit `self.`?

File: NimbleMedium/Sources/Supports/Extensions/String.swift [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+//  String.swift
+//  NimbleMedium
+//
+//  Created by Mark G on 10/08/2021.
+//
+
+extension String {
+
+    static let empty: String = ""
+}
```
Comment (suho) [inline NimbleMedium/Sources/Supports/Extensions/String.swift:11]: We should move this into the `Constants` folder.

Comment (suho) [reply NimbleMedium/Sources/Supports/Extensions/String.swift:11]: And update the name of this file and move to Foundation folder: `Extensions/Foundation/Optional+String.swift`

File: NimbleMedium/Sources/Supports/Extensions/NavigationBarModifier.swift [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+            }
+        }
+    }
+}
+
+extension View {
+
+    func navigationBarColor(
+        backgroundColor: Color = .clear,
+        titleColor: Color = .white) -> some View {
```
Comment (suho) [inline NimbleMedium/Sources/Supports/Extensions/NavigationBarModifier.swift:53]:
```suggestion
    func navigationBarColor(
        backgroundColor: Color = .clear,
        titleColor: Color = .white
     ) -> some View {
```

File: NimbleMedium/Sources/Presentation/Views/Screen.swift (hunk 5 above)
Comment (markgravity) [reply NimbleMedium/Sources/Presentation/Views/Screen.swift:12]: @suho  Hmm I think we should use it correctly, for example
```swift
struct HomeView: View {

    private let content: () -> Content
    @State var buttonTitle: String = "Append me"

    var body: some View {
        print("Rebuild")

        // TODO: Update the title and its content in Home integrate task
        return Screen(title: Localizable.homeTitle()) {
            content()
            Button {
                buttonTitle += "1"
            } label: {
                Text(buttonTitle)
            }

        }
    }
}
```
In this case, every time we click the button, the body will be recalled. Then if we use `content()` here, it will be re-created also

File: NimbleMedium/Sources/Supports/Extensions/String.swift (hunk 8 above)
Comment (markgravity) [reply NimbleMedium/Sources/Supports/Extensions/String.swift:11]: What kind of extension that we should put into `Foundation`? 🤔

File: NimbleMedium/Sources/Presentation/Views/Screen.swift (hunk 5 above)
Comment (suho) [reply NimbleMedium/Sources/Presentation/Views/Screen.swift:12]: I think It's normal.
From this link: https://www.hackingwithswift.com/books/ios-swiftui/custom-containers. It defines `let content: (Int, Int) -> Content`.

also, with this code, another views will be re-created, right?
```swift
var body: some View {
       Screen(title: Localizable.homeTitle()) {
            content()
            Button {
                buttonTitle += "1"
            } label: {
                Text(buttonTitle) // <--- everytime we press the button, it will re-created the Text?
            }
            Text("Another Text") // <--- it will re-created the Text?
            HStack {} // <--- it will re-created the HStack?
        }
    }
```

Correct me if I'm missing something :D

Comment (suho) [reply NimbleMedium/Sources/Presentation/Views/Screen.swift:12]: @markgravity However, I think we can go with both ways. In some cases like the GridStack in the article, we can use `let content: (Int, Int) -> Content`, otherwise, we can use `let content: Content`
Not a big deal here :D

**Comment by minhnimble** [comment]
Need to fix CI failed issue 🦀🛠

**Comment by minhnimble** [comment]
@markgravity I don't see any CI check after you force-push your branch. Your claim [here](https://github.com/nimblehq/nimble-medium-ios/pull/14#discussion_r684963784) doesn't seem to be correct 😔. We need to add `push` events for our CI, I will reopen the setup CI ticket 🙏

**Review by minhnimble — APPROVED** [review]
LGTM 🚀

**Comment by markgravity** [comment]
@suho @minhnimble I did a bit of refactoring to make our navigation bar configure easier and more readable. Please help me to re-review it 🙏 
And sorry for this re-request, I did it after learning something new in SwiftUI 😁 

Before
```swift
Screen(title: "Home") {
    Text("This home screen")
        .toolbar {
            // Add more navigation items here
        }
}
```

After
```swift
Screen() {
    NavigationBar (title: "Home") {
        // Add more navigation items here
    }
    Text("This home screen")
}
```

**Review by minhnimble — APPROVED** [review]
> @suho @minhnimble I did a bit of refactoring to make our navigation bar configure easier and more readable. Please help me to re-review it 🙏
> And sorry for this re-request, I did it after learning something new in SwiftUI 😁
> 
> Before
> 
> ```swift
> Screen(title: "Home") {
>     Text("This home screen")
>         .toolbar {
>             // Add more navigation items here
>         }
> }
> ```
> 
> After
> 
> ```swift
> Screen {
>     NavigationBar(title: "Home") {
>         // Add more navigation items here
>     }
>     Text("This home screen")
> }
> ```

@markgravity don't worry we are all learning, nice work 👍 Just a friendly reminder that you might need to update the P.o.W if needed since we are changing the way to create the navigation bar a bit🙏

**Comment by suho** [comment]
> @suho @minhnimble I did a bit of refactoring to make our navigation bar configure easier and more readable. Please help me to re-review it 🙏
> And sorry for this re-request, I did it after learning something new in SwiftUI 😁
> 
> Before
> 
> ```swift
> Screen(title: "Home") {
>     Text("This home screen")
>         .toolbar {
>             // Add more navigation items here
>         }
> }
> ```
> 
> After
> 
> ```swift
> Screen {
>     NavigationBar(title: "Home") {
>         // Add more navigation items here
>     }
>     Text("This home screen")
> }
> ```

I would not prefer this way, because you are not familiar with declarative syntax. SwiftUI uses a declarative syntax, so it's fine to have this code:
```swift
Text("This home screen")
    .toolbar {  ... }
// this "toolbar" way is easier and readable for people who already knew SwiftUI :D.
```
If the codes are not easier or readable for you, so you need to learn and understand more about **declarative syntax**.
Don't do something unusual this time, new SwiftUI members (or even people who have experience with SwiftUI) when joining the project will take time to check the source code to understand. We can do some customize later after all members are familiar with SwiftUI and we have a convention with SwiftUI.

In this project, let's keep the SwiftUI style first, don't try to customize it. We need to understand deeply the SwiftUI first.

In conclusion, I prefer to use `NavigationView`, `.navigationBarTitle`, `. navigationBarColor `, `. toolbar` directly in the `HomeView`. We don't need `Screen` and `NavigationBar` at this time :D

**Review by suho — CHANGES_REQUESTED** [review]
Please check my comment above ☝️

**Comment by markgravity** [comment]
@suho 
> I would not prefer this way, because you are not familiar with the declarative syntax, we need to learn it. SwiftUI uses a declarative syntax, so it's fine to have this code:

I don't mean the syntax there, the meaning I wanna mention here is that when we use a view modifer you will think it will modify that view.
But with `toolbar`, it modifies the navigation bar, but it's called from `Text` that it makes me look weird

> In this project, let's keep the SwiftUI style first, don't try to customize it. We need to understand deeply the SwiftUI first.

If we agree to stick with the original stuff, let me revert these 😢 
cc: @minhnimble

**Comment by suho** [comment]
> but it's called from Text that it makes me look weird

@markgravity That's why we need to learn SwiftUI to understand more. Why did Apple put the toolbar like that, why don't they implement SwiftUI like your way :D SwiftUI also has `NavigationLink`, you will also put it into the content block of NavigationView. So lot of new thing to learn, don't try to ignore it :D

**Review by suho — APPROVED** [review]
